src/stan/variational/advi.hpp

```diff
@@ -522,36 +381,369 @@ class advi {
+      hist_vector.push_back(histMat(n_approx_params, max_runs));
+      variational_obj_vec.push_back(Q(cont_params_));
+      elbo_grad_vec.push_back(Q(dim));
+      elbo_grad_square_vec.push_back(Q(dim));
+    }
+
+    for (int n_iter = 0; n_iter < max_runs; n_iter++){
+      eta_scaled = eta / sqrt(static_cast<double>(n_iter + 1));
+      for (int n_chain = 0; n_chain < num_chains; n_chain++){
+        calc_ELBO_grad(variational_obj_vec[n_chain], elbo_grad_vec[n_chain], logger);
+        if (n_iter == 0) {
+          elbo_grad_square_vec[n_chain] += elbo_grad_vec[n_chain].square();
+        }
+        else {
+          elbo_grad_square_vec[n_chain] = 0.9 * elbo_grad_square_vec[n_chain]
+                                          + 0.1 * elbo_grad_vec[n_chain].square();
+        }
+        variational_obj_vec[n_chain] += eta_scaled * elbo_grad_vec[n_chain] / (1.0 + elbo_grad_square_vec[n_chain].sqrt());
+
+        hist_vector[n_chain].col(n_iter) = variational_obj_vec[n_chain].return_approx_params();
+      }
+
+      if ((n_iter % eval_window == 0 && n_iter > 0) || n_iter == max_runs - 1){
+        max_rhat = std::numeric_limits<double>::lowest();
+        for(int k = 0; k < n_approx_params; k++) {
+          std::vector<const double*> hist_ptrs;
+          std::vector<size_t> chain_length;
+          if(num_chains == 1){
+            // use split rhat
+            chain_length.assign(2, static_cast<size_t>(n_iter/2 * window_size));
+            hist_ptrs.push_back(hist_vector[0].row(k).data());
+            hist_ptrs.push_back(hist_ptrs[0] + chain_length[0]);
+          }
+          else{
+            for(int i = 0; i < num_chains; i++){
+              //chain_length.push_back(static_cast<size_t>(n_iter * window_size));
+              //hist_ptrs.push_back(hist_vector[i].row(k).data());
+
+              // multi-chain split rhat (split each chain into 2)
+              chain_length.insert(chain_length.end(), 2, static_cast<size_t>(n_iter/2 * window_size));
+              hist_ptrs.push_back(hist_vector[i].row(k).data());
+              hist_ptrs.push_back(hist_vector[i].row(k).data() + chain_length[0]);
+            }
+          }
+          rhat = stan::analyze::compute_potential_scale_reduction(hist_ptrs, chain_length);
+          max_rhat = std::max<double>(max_rhat, rhat);
+        }
+
+        if (max_rhat < rhat_cut) {
+          T0 = n_iter;
+          ss << "Preliminary iterations terminated by rhat condition at iteration # " << T0 <<
+                " with max reported Rhat value of " << max_rhat << "\n";
+          break;
+        }
+      }
+    }
+
+    ss << "Pre-iteration done. T0: " << T0 << "\n";
+
+    bool khat_failed = false;
+    for(int k = 0; k < num_chains; k++){
+      Eigen::VectorXd lw_vec(n_posterior_samples_);
+      lr(variational_obj_vec[k], lw_vec);
+      double sigma, max_lw;
+      int n_tail;
+      if(n_posterior_samples_ < 225) {
+        n_tail = int(lw_vec.size() * 0.2);
```

Comment:
@hyunjimoon 
I'm not sure how I should change this.
[loo](https://github.com/stan-dev/loo/blob/51fb2d0511a765407ff1bd47fab7e79b5c62cbfb/R/psis.R#L276) uses `3 * sqrt(N / ess)`